FastSimulation/Tracking/src/SeedFinderSelector.cc

```diff
@@ -119,8 +148,88 @@ bool SeedFinderSelector::pass(const std::vector<const FastTrackerRecHit *>& hits
 	    multiHitGenerator_->hitTriplets(*trackingRegion_,tripletresult,*eventSetup_,result,&thmp,thirdLayerDetLayer,1);
 	    return !tripletresult.empty();
 	}
-
+	else if(CAHitTriplGenerator_)
+	{  
+	    return true;
+	}
     }
+    
+    if(CAHitQuadGenerator_)
+    {
+      if(hits.size() < 4)
+	{
+	  throw cms::Exception("FastSimTracking") << "For the given configuration, SeedFinderSelector::pass requires at least 4 hits";
+	}
+
+      if(!seedingLayer)
+	throw cms::Exception("FastSimTracking") << "ERROR: SeedingLayers pointer not set";      
+
+      SeedingLayerSetsHits & layers = *seedingLayer;
+      IntermediateHitDoublets ihd(&layers);
+      const TrackingRegion& tr_ = *trackingRegion_;
+      auto filler = ihd.beginRegion(&tr_);
+      
+      std::array<SeedingLayerSetsBuilder::SeedingLayerId,2> hitPair;
+      for(int i=0; i<3; i++){
+	SeedingLayerSetsHits::SeedingLayerSet pairCandidate;
+	hitPair[0] = Layer_tuple(hits[i]);
+ 	hitPair[1] = Layer_tuple(hits[i+1]);
+       
+	bool found;
+        for(SeedingLayerSetsHits::SeedingLayerSet ls : *seedingLayer){
+	  found = false;
+	  for(const auto p : layerPairs_){
+	    pairCandidate = ls.slice(p,p+2);
+	    if(hitPair[0] == seedingLayerIds[pairCandidate[0].index()] && hitPair[1] == seedingLayerIds[pairCandidate[1].index()]){
+	      found = true;
+	      break;
+	    }
+	  }
+	  if(found)
+	    break;
```

Comment:
This `if(found) break;` was what I found to cure the problem you described in https://github.com/angirar/cmssw/pull/2#discussion_r179571227 (thanks for the recipe). With it I got CA to enter to the loop
https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc#L29
but there are still now output quadruplets. I poked around a bit, and the reason seems to lie in the `haveSimilarCurvature()` call in
https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/CACell.h#L143
i.e. it returns always `false` (in the 10 events). I find this a bit strange (could there be a systematic "shift" somewhere to make the doublet compatibility check fail?), but maybe you can continue the investigation from here? (for further help in there you/we'll may need Felice as well).

Reply:
And actually what I had originally in mind was adding a sanity check/assert after the outermost for-loop that some element was indeed found.

Reply:
I have actually verified for 10 events that a pairCandidate is always found and I have put a std::assert() statement also. I don't think that doublet compatibility check is failing anywhere (since the IHD are being filled nicely) but again, I'm not so sure. I'll try to investigate more starting from the function haveSimilarCurvature(). But it might help to bring in more help from Felice sooner rather than later.

Reply:
I looked into the haveSimilarCurvature() to see why it returns false everytime. It appears that the actual problem lies with the 'doublets' that are stored into std::vector of CACell "allCells" by CellularAutomaton.cc at https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/CellularAutomaton.cc#L63.
I think so because certain parameters (X,Y,Z,R,Phi) are extracted from these doublets (https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/CACell.h#L45) which is then used in multiple places but all of them return values 0. As a result of this, haveSimilarCurvature() fails the condition here https://github.com/cms-sw/cmssw/blob/master/RecoPixelVertexing/PixelTriplets/plugins/CACell.h#L219 (distance_13_beamspot_squared = -nan). 
Do you have any idea why the doublets would be improper?

Reply:
Any insights here?

Reply:
Not yet... but I'm taking a look.

Reply:
The problem is that below lines 202-204
```cpp
const RecHitsSortedInPhi firsthm(fHits, trackingRegion_->origin(), fLayer);
const RecHitsSortedInPhi secondhm(sHits, trackingRegion_->origin(), sLayer);
HitDoublets res(firsthm,secondhm);
```
do not work as intended as `HitDoublets` stores the pointers to the input objects and assumes they are managed (owned) somebody else (in standard seeding this is `LayerHitMapCache`). In this loop, however, the `RecHitsSortedInPhi` are temporary and thus `HitDoublets` eventually has pointers to "random memory".

I managed to get a step forward (if not finally fixing it) by replacing lines 202-203 with
```
auto& layerCache = filler.layerHitMapCache();
const RecHitsSortedInPhi& firsthm = *layerCache.add(pairCandidate[0], std::make_unique<RecHitsSortedInPhi>(fHits, trackingRegion_->origin(), fLayer));
const RecHitsSortedInPhi& secondhm = *layerCache.add(pairCandidate[1], std::make_unique<RecHitsSortedInPhi>(sHits, trackingRegion_->origin(), sLayer));
```
and making the following modifications to `LayerHitMapCache.h`
https://github.com/cms-sw/cmssw/commit/92a3f2ca2e241f7797000d9340128e0f9cc66a7c

Could you try if those would solve the problem for good?